lib/debugger/debugger.dart

```diff
@@ -0,0 +1,1309 @@
+    // AsyncCausal frames don't have local vars.
+    _items.setItems(frame.vars ?? <BoundVariable>[]);
+  }
+
+  void clearVariables() {
+    _items.setItems(<BoundVariable>[]);
+  }
+}
+
+class BreakOnExceptionControl extends CoreElement {
+  BreakOnExceptionControl() : super('div', classes: 'break-on-exceptions') {
+    final CoreElement unhandled = new CoreElement('input')
+      ..setAttribute('type', 'checkbox');
+    _unhandledElement = unhandled.element;
+
+    final CoreElement all = new CoreElement('input')
+      ..setAttribute('type', 'checkbox');
+    _allElement = all.element;
+
+    add([
+      span(text: 'Break on: '),
+      new CoreElement('label')
+        ..add(<CoreElement>[unhandled, span(text: ' Unhandled exceptions')]),
+      new CoreElement('label')
+        ..add(<CoreElement>[all, span(text: ' All exceptions')]),
+    ]);
+
+    unhandled.element.onChange.listen((_) {
+      _pauseModeController.add(exceptionPauseMode);
+    });
+
+    all.element.onChange.listen((_) {
+      if (_allElement.checked) {
+        unhandled.enabled = false;
+        _unhandledElement.checked = true;
+      } else {
+        unhandled.enabled = true;
+      }
+      _pauseModeController.add(exceptionPauseMode);
+    });
+  }
+
+  html.InputElement _unhandledElement;
+  html.InputElement _allElement;
+
+  final StreamController<String> _pauseModeController =
+      new StreamController.broadcast();
+
+  /// See the string values for [ExceptionPauseMode].
+  Stream<String> get onPauseModeChanged => _pauseModeController.stream;
+
+  String get exceptionPauseMode {
+    if (_allElement.checked) {
+      return ExceptionPauseMode.kAll;
+    } else if (_unhandledElement.checked) {
+      return ExceptionPauseMode.kUnhandled;
+    } else {
+      return ExceptionPauseMode.kNone;
+    }
+  }
+
+  set exceptionPauseMode(final String value) {
+    if (value == ExceptionPauseMode.kAll) {
+      _allElement.checked = true;
+      _unhandledElement.checked = true;
+      _unhandledElement.setAttribute('disabled', '');
+    } else if (value == ExceptionPauseMode.kUnhandled) {
+      _allElement.checked = false;
+      _unhandledElement.checked = true;
+      _unhandledElement.attributes.remove('disabled');
+    } else {
+      _allElement.checked = false;
+      _unhandledElement.checked = false;
+      _unhandledElement.attributes.remove('disabled');
+    }
+  }
+}
+
+class ScriptAndPos {
+  ScriptAndPos(this.uri, {this.position});
+
+  final String uri;
+  final Pos position;
+
+  bool matches(Script script) => uri == script.uri;
+}
+
+int _breakpointComparator(Breakpoint a, Breakpoint b) {
+  ScriptRef getRef(dynamic location) {
+    if (location is SourceLocation) {
+      return location.script;
+    } else if (location is UnresolvedSourceLocation) {
+      return location.script;
+    } else {
+      return null;
+    }
+  }
+
+  int getPos(dynamic location) {
+    if (location is SourceLocation) {
+      return location.tokenPos ?? 0;
```

Comment:
nit: make location.tokenPos be made non-nullable instead of having a check here?

Reply:
We get this info directly from the service protocol - in some cases tokenPos can be null.